File: app/src/main/java/me/eycia/api/MsgBase.java
package me.eycia.api;

import android.os.Parcel;
import android.os.Parcelable;

/**
 * Created by eycia on 2/28/16.
 */
public class MsgBase implements Parcelable {
    public String Id;
    public long SnapTime;
    public long PubTime;
    public String SourceURL;
    public String Title;
    public String SubTitle;
    public String CoverImg;   //length zero if not have
    public int ViewType;
    public String AuthorId;
    public String AuthorCoverImg;
    public String AuthorName;
    public String Tag;
    public String Topic;        //length zero if not have

    protected MsgBase(Parcel in) {
        Id = in.readString();
        SnapTime = in.readLong();
        PubTime = in.readLong();
        SourceURL = in.readString();
        Title = in.readString();
        SubTitle = in.readString();
        CoverImg = in.readString();
        ViewType = in.readInt();
        AuthorId = in.readString();
        AuthorCoverImg = in.readString();
        AuthorName = in.readString();
        Tag = in.readString();
        Topic = in.readString();
    }

    public static final Creator<MsgBase> CREATOR = new Creator<MsgBase>() {
        @Override
        public MsgBase createFromParcel(Parcel in) {
            return new MsgBase(in);
        }

        @Override
        public MsgBase[] newArray(int size) {
            return new MsgBase[size];
        }
    };

    public MsgBase(String authorCoverImg, String authorId, String authorName, String coverImg, String id, long pubTime, long snapTime, String sourceURL, String subTitle, String tag, String title, String topic, int viewType) {
        AuthorCoverImg = authorCoverImg;
        AuthorId = authorId;
        AuthorName = authorName;
        CoverImg = coverImg;
        Id = id;
        PubTime = pubTime;
        SnapTime = snapTime;
        SourceURL = sourceURL;
        SubTitle = subTitle;
        Tag = tag;
        Title = title;
        Topic = topic;
        ViewType = viewType;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(Id);
        dest.writeLong(SnapTime);
        dest.writeLong(PubTime);
        dest.writeString(SourceURL);
        dest.writeString(Title);
        dest.writeString(SubTitle);
        dest.writeString(CoverImg);
        dest.writeInt(ViewType);
        dest.writeString(AuthorId);
        dest.writeString(AuthorCoverImg);
        dest.writeString(AuthorName);
        dest.writeString(Tag);
        dest.writeString(Topic);
    }
}

File: app/src/main/java/me/eycia/api/MsgBase.kt
package me.eycia.api

import android.os.Parcel
import android.os.Parcelable

data class MsgBase(var Id: String,
                   var Body: String, //length zero if not have
                   var SnapTime: Long,
                   var PubTime: Long,
                   var SourceURL: String,
                   var Title: String,
                   var SubTitle: String,
                   var CoverImg: String, //length zero if not have
                   var ViewType: Int,
                   var AuthorId: String,
                   var AuthorCoverImg: String,
                   var AuthorName: String,
                   var Tag: String,
                   var Topic: String, //length zero if not have
                   var PicRefs: Array<PicRef>
) : Parcelable, Comparable<MsgBase> {

    override fun compareTo(other: MsgBase): Int {
        return other.SnapTime.compareTo(SnapTime)
    }

    constructor(`in`: Parcel) : this(
            Id = `in`.readString(),
            Body = `in`.readString(),
            SnapTime = `in`.readLong(),
            PubTime = `in`.readLong(),
            SourceURL = `in`.readString(),
            Title = `in`.readString(),
            SubTitle = `in`.readString(),
            CoverImg = `in`.readString(),
            ViewType = `in`.readInt(),
            AuthorId = `in`.readString(),
            AuthorCoverImg = `in`.readString(),
            AuthorName = `in`.readString(),
            Tag = `in`.readString(),
            Topic = `in`.readString(),
            PicRefs = `in`.createTypedArray(PicRef.CREATOR)) {
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(Id)
        dest.writeString(Body)
        dest.writeLong(SnapTime)
        dest.writeLong(PubTime)
        dest.writeString(SourceURL)
        dest.writeString(Title)
        dest.writeString(SubTitle)
        dest.writeString(CoverImg)
        dest.writeInt(ViewType)
        dest.writeString(AuthorId)
        dest.writeString(AuthorCoverImg)
        dest.writeString(AuthorName)
        dest.writeString(Tag)
        dest.writeString(Topic)
        dest.writeTypedArray(PicRefs, flags)
    }

    companion object {

        @JvmField final val CREATOR: Parcelable.Creator<MsgBase> = object : Parcelable.Creator<MsgBase> {
            override fun createFromParcel(`in`: Parcel): MsgBase {
                return MsgBase(`in`)
            }

            override fun newArray(size: Int): Array<MsgBase?> {
                return arrayOfNulls(size)
            }
        }
    }
}
